### PR TITLE
[mlir][debug] Allow multiple DIGlobalVariableExpression on globals.

### PR DESCRIPTION
**Pull request by abidh**
Currently, we allow only one  DIGlobalVariableExpressionAttr per global. It is especially evident in import where we pick the first from the list and ignore the rest. In contrast, LLVM allows multiple DIGlobalVariableExpression to be attached to the global. They are needed for correct working of things like DICommonBlock. This PR removes this restriction in mlir. Changes are mostly mechanical. One thing on which I went a bit back and forth was the representation inside GlobalOp. I would be happy to change if there are better ways to do this.

### CODE REVIEW

File: mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp [hunk 1]
```diff
@@ -2089,7 +2089,7 @@ void GlobalOp::build(OpBuilder &builder, OperationState &result, Type type,
                      Attribute value, uint64_t alignment, unsigned addrSpace,
                      bool dsoLocal, bool threadLocal, SymbolRefAttr comdat,
                      ArrayRef<NamedAttribute> attrs,
-                     DIGlobalVariableExpressionAttr dbgExpr) {
+                     ArrayRef<Attribute> dbgExprs) {
```
Comment (gysit) [inline mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp:2092]:
```suggestion
                     ArrayRef<DIGlobalVariableExpressionAttr> dbgExprs) {
```
Can this be typed (here and in the other places)?

File: mlir/lib/Target/LLVMIR/ModuleTranslation.cpp [hunk 2]
```diff
@@ -1055,44 +1055,50 @@ LogicalResult ModuleTranslation::convertGlobals() {
...
-      // Get the compile unit (scope) of the the global variable.
-      if (llvm::DICompileUnit *compileUnit =
-              dyn_cast_if_present<llvm::DICompileUnit>(scope)) {
-        // Update the compile unit with this incoming global variable expression
-        // during the finalizing step later.
-        allGVars[compileUnit].push_back(diGlobalExpr);
+    if (op.getDbgExprs()) {
+      for (auto attr : *op.getDbgExprs()) {
+        if (auto exprAttr =
+                dyn_cast_if_present<DIGlobalVariableExpressionAttr>(attr)) {
```
Comment (gysit) [inline mlir/lib/Target/LLVMIR/ModuleTranslation.cpp:1061]: Can the debug expression array ever have a nullptr entry or contain something else than a `DIGlobalVariableExpressionAttr`? I would presume not and a `cast<DIGlobalVariableExpressionAttr>(attr) `makes more range. Maybe you can also use `op.getDbgExprs().getAsRange<DIGlobalVariableExpressionAttr>()`?

File: mlir/test/Target/LLVMIR/llvmir-debug.mlir [hunk 3]
```diff
@@ -690,3 +690,27 @@ llvm.func @test() {
 // CHECK: !DICommonBlock(scope: ![[SCOPE:[0-9]+]], declaration: null, name: "block", file: ![[FILE:[0-9]+]], line: 3)
 // CHECK: ![[SCOPE]] = {{.*}}!DISubprogram(name: "test"{{.*}})
 // CHECK: ![[FILE]] = !DIFile(filename: "test.f90"{{.*}})
+
+// -----
+
+// Test multiple DIGlobalVariableExpression on a global
```
Comment (gysit) [inline mlir/test/Target/LLVMIR/llvmir-debug.mlir:696]:
```suggestion
// Test multiple DIGlobalVariableExpression on a global.
```
ultra nit:

File: mlir/include/mlir/Dialect/LLVMIR/LLVMOps.td [hunk 4]
```diff
@@ -1152,6 +1153,9 @@ def LLVM_AddressOfOp : LLVM_Op<"mlir.addressof",
   let hasFolder = 1;
 }
 
+def DIGlobalVariableExpressionArrayAttr :
+  TypedArrayAttrBase<LLVM_DIGlobalVariableExpressionAttr, "an array of variable expressions">;
```
Comment (gysit) [inline mlir/include/mlir/Dialect/LLVMIR/LLVMOps.td:1157]: I would consider moving this right below the definition of LLVM_DIGlobalVariableExpressionAttr? We already have such arrays for example for LLVM_AliasScopeArrayAttr.

**Review by gysit — COMMENTED** [review]
Thanks! I added some nit comments.

I think the approach with a typed array attribute makes sense. There is already prior art for the alias analysis metadata. 

What alternative did you have in mind?

File: mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp (hunk 1 above)
Comment (abidh) [reply mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp:2092]: I tried this but it did not work. The `DIGlobalVariableExpressionArrayAttr` becomes an ArrayAttr and that takes an `ArrayRef<Attribute>` and compiler does not allow `ArrayRef<DIGlobalVariableExpressionAttr>` to be passed for that. I thought doing manual conversion was not worth the gain.

**Comment by abidh** [comment]
> Thanks! I added some nit comments.
> 
> I think the approach with a typed array attribute makes sense. There is already prior art for the alias analysis metadata.
> 
> What alternative did you have in mind?

As I mentioned in the comment above, I was thinking if I could make the argument to the build function more strongly typed. But using `ArrayRef<DIGlobalVariableExpressionAttr>` did not work out.

File: mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp (hunk 1 above)
Comment (gysit) [reply mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp:2092]: Ok makes sense! I think DIGlobalVariableExpressionArrayAttr has a verifier so things should ultimately be checked.

File: mlir/lib/Target/LLVMIR/ModuleTranslation.cpp (hunk 2 above)
Comment (abidh) [reply mlir/lib/Target/LLVMIR/ModuleTranslation.cpp]: Thanks for the suggestion. I have replaced it with `op.getDbgExprs()->getAsRange<DIGlobalVariableExpressionAttr>()`.

File: mlir/test/Target/LLVMIR/llvmir-debug.mlir (hunk 3 above)
Comment (abidh) [reply mlir/test/Target/LLVMIR/llvmir-debug.mlir]: Thanks for catching this. Fixed.

File: mlir/include/mlir/Dialect/LLVMIR/LLVMOps.td (hunk 4 above)
Comment (abidh) [reply mlir/include/mlir/Dialect/LLVMIR/LLVMOps.td]: Done.

File: mlir/lib/Target/LLVMIR/ModuleImport.cpp [hunk 5]
```diff
@@ -914,14 +914,16 @@ LogicalResult ModuleImport::convertGlobal(llvm::GlobalVariable *globalVar) {
...
-  DIGlobalVariableExpressionAttr globalExpressionAttr;
+  SmallVector<Attribute> globalExpressionAttrs;
   SmallVector<llvm::DIGlobalVariableExpression *> globalExpressions;
   globalVar->getDebugInfo(globalExpressions);
 
   // There should only be a single global expression.
-  if (!globalExpressions.empty())
-    globalExpressionAttr =
-        debugImporter->translateGlobalVariableExpression(globalExpressions[0]);
+  for (auto expr : globalExpressions) {
```
Comment (gysit) [inline mlir/lib/Target/LLVMIR/ModuleImport.cpp:922]:
```suggestion
  for (Attribute expr : globalExpressions) {
```
nit:

**Review by gysit — APPROVED** [review]
Thanks LGTM modulo nit!

**Pull request merged**